samples: SPI sample (#761)
nrfconnect/sdk-nrf-bm: samples/peripherals/spi/README.rst

********

The sample initializes the **SPIM** and **SPIS** instances with the pins configured in the board's :file:`board-config.h`.
When **Button 2** is pressed, the master sends a configurable string (:kconfig:option:`CONFIG_SAMPLE_SPI_MSG`) to the slave.
The slave receives the data, logs it, and toggles **LED 2**.

User interface
**************

LED 0:
  Lit when the sample is initialized and ready.

LED 2:
  Toggles each time the SPI slave completes a reception.

Button 2:
  Sends a message from the local SPIM to the connected SPIS.

Building and running
********************

This sample can be found under :file:`samples/peripherals/spi/` in the |BMshort| folder structure.

For details on how to create, configure, and program a sample, see :ref:`getting_started_with_the_samples`.

Testing
=======

You can test this sample by performing the following steps:

#. Compile and program the application.
#. Open the console log and verify that ``SPI sample initialized`` appears in the log.
#. Press **Button 2**.
#. Observe that the log shows a master "sent" entry and a slave "received" entry, both containing the string set by the :kconfig:option:`CONFIG_SAMPLE_SPI_MSG` Kconfig option.

> #. Observe that **LED 2** toggles on each completed reception.

Reviewer: This could use some updates to facilitate for the two-device test-case.

SPI
###

.. contents::
   :local:
   :depth: 2

The SPI sample demonstrates how to configure and use the SPIM and SPIS peripherals with the nrfx drivers.

Requirements
************

The sample supports the following development kits:

.. tabs::

   .. group-tab:: Simple board variants

      The following board variants do **not** have DFU capabilities:

      .. include:: /includes/supported_boards_all_non-mcuboot_variants_s115.txt

      .. include:: /includes/supported_boards_all_non-mcuboot_variants_s145.txt

   .. group-tab:: MCUboot board variants

      The following board variants have DFU capabilities:

      .. include:: /includes/supported_boards_all_mcuboot_variants_s115.txt

      .. include:: /includes/supported_boards_all_mcuboot_variants_s145.txt

Overview
********

The sample initializes the **SPIM** and **SPIS** instances with the pins configured in the :file:`board-config.h` header.
When **Button 2** is pressed, a configurable string (:kconfig:option:`CONFIG_SAMPLE_SPI_MSG`) is sent from the local SPI controller to a connected SPI target.
The local SPI target receives data from a connected SPI controller, logs the received data and toggles **LED 2**.

Wiring

> *******

Reviewer: ```suggestion
Wiring
******
```
Fix this if updating other stuff.

Author: Good catch, fixed

Button 2:
   Send string from the local SPIM to a connected SPI target.

Configuration
*************

You can modify the following options (available in the Kconfig file at :file:`samples/peripherals/spi`):

* :kconfig:option:`CONFIG_SAMPLE_SPI_MSG` - Set the message that the SPI controller sends to the SPI target when **Button 2** is pressed.

* :kconfig:option:`CONFIG_SAMPLE_SPI_RX_BUF_SIZE` - Set the size, in bytes, of the buffer used by the SPI target to receive data.

* :kconfig:option:`CONFIG_SAMPLE_SPI_CSN_TO_CLK_DELAY` - Set the delay, in SPIM core clock periods, between chip select (CSN) going active and the first SCK edge.
  The required delay depends on how long the connected SPI target takes to wake up and become ready to receive data, which varies between devices.
  Refer to the electrical specification chapter of the SPI target device's datasheet for its wake-up time.

Building and running
********************

This sample can be found under :file:`samples/peripherals/spi/` in the |BMshort| folder structure.

For details on how to create, configure, and program a sample, see :ref:`getting_started_with_the_samples`.

Testing
=======

> You can test the sample in two ways, depending on the selected wiring: the single-device loopback setup or the two-device setup.

Reviewer: ```suggestion
You can test the sample in two ways depending on the selected wiring:
```

The tab titles are visible below, so I don't think we need to repeat here.

Author: I am not sure if we should remove this. I tried to align README with how it is done other places where 2 testing ways are possible (for example `ble_nus`). There we state both options:
https://github.com/nrfconnect/sdk-nrf-bm/blob/main/samples/bluetooth/ble_nus/README.rst?plain=1#L75

.. _bm_spi_sample:

SPI
###

.. contents::
   :local:
   :depth: 2

The SPI sample demonstrates how to configure and use the SPIM and SPIS peripherals with the nrfx drivers.

Requirements
************

The sample supports the following development kits:

.. tabs::

>    .. group-tab:: Simple board variants

Reviewer 2: Use templates for this

Author: fixed

         * - nRF54LS05 DK
           - PCA10214
           - S115
           - bm_nrf54ls05dk/nrf54ls05b/cpuapp/s115_softdevice/mcuboot
         * - `nRF54LV10 DK`_
           - PCA10188
           - S115
           - bm_nrf54lv10dk/nrf54lv10a/cpuapp/s115_softdevice/mcuboot
         * - `nRF54L15 DK`_
           - PCA10156
           - S145
           - bm_nrf54l15dk/nrf54l15/cpuapp/s145_softdevice/mcuboot
         * - `nRF54L15 DK`_ (emulating nRF54L10)
           - PCA10156
           - S145
           - bm_nrf54l15dk/nrf54l10/cpuapp/s145_softdevice/mcuboot
         * - `nRF54L15 DK`_ (emulating nRF54L05)
           - PCA10156
           - S145
           - bm_nrf54l15dk/nrf54l05/cpuapp/s145_softdevice/mcuboot
         * - `nRF54LM20 DK`_
           - PCA10184
           - S145
           - bm_nrf54lm20dk/nrf54lm20a/cpuapp/s145_softdevice/mcuboot
         * - nRF54LS05 DK
           - PCA10214
           - S145
           - bm_nrf54ls05dk/nrf54ls05b/cpuapp/s145_softdevice/mcuboot
         * - `nRF54LV10 DK`_
           - PCA10188
           - S145
           - bm_nrf54lv10dk/nrf54lv10a/cpuapp/s145_softdevice/mcuboot

.. _bm_spi_wiring:

> Wiring

Reviewer 2: This belongs under test step

Author: fixed

======

The sample uses the **SPIM** and **SPIS** pins defined in the board's :file:`board-config.h` header (``BOARD_APP_SPIM_*`` and ``BOARD_APP_SPIS_*``).

.. include:: /includes/spi_board_connections_nrf54l15.txt

Connect the pins in one of the following configurations:

.. tabs::

   .. tab:: Single device (loopback)

      Connect the four SPIM pins to the four SPIS pins **on the same board**, matching each signal by name (SCK→SCK, MOSI→MOSI, MISO→MISO, CSN→CSN).

   .. tab:: Two devices, one data path

      Wire **Device 1 SPIM** to **Device 2 SPIS** — four signal wires plus GND:

      =================  =================
      Device 1 (SPIM)    Device 2 (SPIS)
      =================  =================
      SCK                SCK
      MOSI               MOSI
      MISO               MISO
      CSN                CSN
      **GND**            **GND**

>       =================  =================

Reviewer 2: Use same formatting as other tables

Author: fixed

      Only Device 1 can send. Press **Button 2** on Device 1 to trigger a transfer.

   .. tab:: Two devices, two data paths

      Wire **both** buses so each device is master toward the other's slave, eight signal wires plus GND:

      .. list-table:: Bus A — Device 1 sends to Device 2
         :widths: auto
         :header-rows: 1

         * - Device 1 (SPIM)
           - Device 2 (SPIS)
         * - SCK
           - SCK
         * - MOSI
           - MOSI
         * - MISO
           - MISO
         * - CSN
           - CSN

      .. list-table:: Bus B — Device 2 sends to Device 1
         :widths: auto
         :header-rows: 1

         * - Device 2 (SPIM)
           - Device 1 (SPIS)
         * - SCK
           - SCK
         * - MOSI
           - MOSI
         * - MISO
           - MISO
         * - CSN
           - CSN

      Connect **GND** between the two devices.

      Press **Button 2** on either device to send to the other.

> .. note:: LED 3 and SPIM CSN (nRF54L15)

Reviewer 2: We should not have four notes after each other.

Author: fixed

   On the `nRF54L15 DK`_, P1.14 is shared between **SPIM CSN** and **LED 3**.
   The peripheral drives chip select, so the LED is not independent and flickers during transfers.

.. note:: Button 0 and SPIM MISO (nRF54L15)

   On the `nRF54L15 DK`_, P1.13 is shared between **SPIM MISO** and **Button 0**.
   Do not press Button 0 while the SPI bus is active.

.. note:: UARTE pin conflict (nRF54L15)

   On the `nRF54L15 DK`_, **SPIS** pins P0.00-P0.03 overlap with the default **Application UARTE** in :file:`board-config.h`.
   The :ref:`uarte_sample` and this sample cannot use those pins at the same time.

>    Reassign one peripheral in the board file if both are needed.

Reviewer 2: This can be removed. The sample does not use application uarte.

Author: fixed

.. note:: LPUARTE pin conflict (nRF54L15)

   On the `nRF54L15 DK`_, **LPUARTE TX, RX, and RDY** share P1.11, P1.12, and P1.14 with **SPIM SCK, MOSI, and CSN**. **LPUARTE REQ** (P0.04) overlaps with **Button 3**.

>    The :ref:`bm_lpuarte_sample` and this sample cannot run together without reassigning pins in the board file.

Reviewer 2: We run one sample at a time. This comment is not needed.

Author: fixed

Optionally, connect a logic analyzer or oscilloscope to the SPI lines.

> Use a short GND lead to the dev kit ground and a sample rate high enough to resolve the 4 MHz default SPI clock.

Reviewer 2: This is part of testing and if kept should be placed there. I think details on testing with logic analyzer is something we can remove.

Author: oki, removed

Overview
********

> The sample initializes the **SPIM** and **SPIS** instances from the board's :file:`board-config.h`.

Reviewer 2: ```suggestion
The sample initializes the **SPIM** and **SPIS** instances with the pins configured in the board's :file:`board-config.h`.
```

Author: fixed

When **Button 2** is pressed, the master sends a configurable string (:kconfig:option:`CONFIG_SAMPLE_SPI_MSG`) to the slave.
The slave receives the data, logs it, and toggles **LED 2**.

User interface
**************

LED 0
   Lit when the sample is initialized and ready.

LED 2
   Toggles each time the SPI slave completes a reception.

Button 2
   Sends a message from the local SPIM to the connected SPIS.

Building and running
********************

> This sample is in :file:`samples/peripherals/spi/` in the |BMshort| folder structure.

Reviewer 2: Let's align with other samples.
```suggestion
This sample can be found under :file:`samples/peripherals/spi/` in the |BMshort| folder structure.
```

Author: fixed

> For general instructions, see :ref:`getting_started_with_the_samples`.

Reviewer 2: ```suggestion
For details on how to create, configure, and program a sample, see :ref:`getting_started_with_the_samples`.
```

Author: fixed

Testing
=======

After building and flashing, verify the wiring configuration you chose:

> .. rubric:: Single device (loopback)

Reviewer 2: We have tabs for this above, let's do it the same way here.

Author: fixed

#. Flash the sample and open a console log.
#. Connect the four SPIM-SPIS signal pairs on the same board (see :ref:`bm_spi_wiring`).
#. Verify ``SPI sample initialized`` appears in the log.
#. Press **Button 2**.
   The log shows a master "sent" entry and a slave "received" entry.
   **LED 2** toggles on each completed reception.

.. rubric:: Two devices, one data path

#. Flash both devices with the same image.
#. Wire the four SPIM→SPIS signals **and GND** between the devices (see :ref:`bm_spi_wiring`).
#. Open a console for each device.
#. Press **Button 2** on Device 1 (the SPIM side).
   Device 1 logs "sent" and Device 2 logs "received".
   **LED 2** on Device 2 toggles.

.. rubric:: Two devices, two data paths

#. Flash both devices with the same image.
#. Wire both sets of four SPIM→SPIS signals (8 wires) **and GND** (see :ref:`bm_spi_wiring`).
#. Open a console for each device.
#. Press **Button 2** on either device to send to the other.
   The sending device logs "sent", the receiving device logs "received",
   and **LED 2** toggles on the receiver.

> A successful test shows the :kconfig:option:`CONFIG_SAMPLE_SPI_MSG` string in both the master completion and slave reception log entries.

Reviewer 2: This is a sample, not a test. 
`Observe ...`

Author: fixed

      .. include:: /includes/supported_boards_all_non-mcuboot_variants_s145.txt

   .. group-tab:: MCUboot board variants

      The following board variants have DFU capabilities:

      .. include:: /includes/supported_boards_all_mcuboot_variants_s115.txt

      .. include:: /includes/supported_boards_all_mcuboot_variants_s145.txt

Overview
********

The sample initializes the **SPIM** and **SPIS** instances with the pins configured in the board's :file:`board-config.h`.
When **Button 2** is pressed, the master sends a configurable string (:kconfig:option:`CONFIG_SAMPLE_SPI_MSG`) to the slave.
The slave receives the data, logs it, and toggles **LED 2**.

User interface
**************

LED 0
  Lit when the sample is initialized and ready.

LED 2
  Toggles each time the SPI slave completes a reception.

Button 2
  Sends a message from the local SPIM to the connected SPIS.

.. _bm_spi_wiring:

Building and running
********************

This sample can be found under :file:`samples/peripherals/spi/` in the |BMshort| folder structure.

For details on how to create, configure, and program a sample, see :ref:`getting_started_with_the_samples`.

Testing
=======

> After building and flashing the sample, test it as follows:

Reviewer: ```suggestion
After building and programming the sample, test it as follows:
```

Author: fixed

1. Wire the four SPIM pins to the SPIS pins on the same board, matching each signal by name (SCK→SCK, MOSI→MOSI, MISO→MISO, CSN→CSN).
   The pins are defined in the board's :file:`board-config.h` header (``BOARD_APP_SPIM_*`` and ``BOARD_APP_SPIS_*``).

   .. include:: /includes/spi_board_connections.txt

   .. note:: Board-specific behavior

      * **nRF54L15 DK** — SPI Master pins overlap with **LED 3** (P1.14, flickers during transfers) and **Button 0** (P1.13, do not press during transfers).
      * **nRF54LV10 DK** — The SPI Slave pins (P0.00-P0.03) are shared with one of the debugger's virtual serial ports.
        Before running the sample, open the `Board Configurator`_ app in `nRF Connect for Desktop`_ and disable the **Connect port VCOM** entry that is mapped to pins **P0.00-P0.03**, to release these pins from the debugger.

>         Leave the other **Connect port VCOM** entry (mapped to pins **P1.04-P1.07**) enabled, since it is the virtual serial port used to read the sample's log output.

Reviewer: Thinking we could move this part containing the wiring setup to the `Requirements` section. After tables with the supported boards. Then explain both the two-device and single-device loopback setup similar to how it has been done in `samples/peripherals/lpuarte/README.rst`.

Author: Agree. I rewrote Testing and Requirements chapter to be better aligned with other samples that test with 2 different setup options, like LPUATE. Take a look and let me know what you think.

When **Button 2** is pressed, the master sends a configurable string (:kconfig:option:`CONFIG_SAMPLE_SPI_MSG`) to the slave.
The slave receives the data, logs it, and toggles **LED 2**.

User interface
**************

LED 0
  Lit when the sample is initialized and ready.

LED 2
  Toggles each time the SPI slave completes a reception.

Button 2
  Sends a message from the local SPIM to the connected SPIS.

.. _bm_spi_wiring:

Wiring
======

The sample uses the **SPIM** and **SPIS** pins defined in the board's :file:`board-config.h` header (``BOARD_APP_SPIM_*`` and ``BOARD_APP_SPIS_*``).

.. include:: /includes/spi_board_connections.txt

Connect the four SPIM pins to the SPIS pins on the same board, matching each signal by name (SCK→SCK, MOSI→MOSI, MISO→MISO, CSN→CSN).

.. note:: Board-specific behavior

   * **nRF54L15 DK** — SPI Master pins overlap with **LED 3** (P1.14, flickers during transfers) and **Button 0** (P1.13, do not press during transfers).
   * **nRF54LV10 DK** — The SPI Slave pins (P0.00-P0.03) are shared with one of the debugger's virtual serial ports.

>      Before running the sample, open the `Board Configurator`_ app in `nRF Connect for Desktop`_ and disable the **Connect port VCOM** entry that is mapped to pins **P0.00-P0.03**, to release these pins from the debugger.

Reviewer 2: Can we use other pins for this? I'd rather not touch the board controller.

Author: I agree this isnt ideal, but its quite tricky. There are no other pins available on Port 0 on the chip itself, nRF54LV10A that sits on the DK only has P0.00-P0.04 on the chip, and thats it.

The alternative would be to move the SPIS from instance 30 to instance 20 on Port 1, which is the other SPIS peripheral instance space available. However, instance 20 on Port 1 is already used by the logger to print output for the user, and we definitely can't touch that. otherwise, the user wouldn't see any output in the logger terminal when running the sample :/

So unfortunately I dont see any less bad solution here. We are kind of stuck with using SPIS30 on P0 if we want both slave and master SPI.

Alternatively master and slave SPI could share the same instance, however that would mean the user requires 2 DKs to test the sample, I don't know if that's something we want either.

Let me know what you think about this situation. If you know how we should proceed with this.

Reviewer 2: Alright, lets keep it as is then.

>      Leave the other **Connect port VCOM** entry (mapped to pins **P1.04-P1.07**) enabled, since it is the virtual serial port used to read the sample's log output.

Reviewer 2: Leave for testing steps.

Author: I moved the connection and wiring to testing chapter. Please check if I did it correctly

Building and running
********************

This sample can be found under :file:`samples/peripherals/spi/` in the |BMshort| folder structure.

For details on how to create, configure, and program a sample, see :ref:`getting_started_with_the_samples`.

Testing
=======

After building and flashing the sample, test it as follows:

1. Connect the four SPIM→SPIS signal pairs on the same board (see :ref:`Wiring <bm_spi_wiring>`).
#. Open the console log and verify that ``SPI sample initialized`` appears in the log.
#. Press **Button 2**.

> #. Observe that the log shows a master "sent" entry and a slave "received" entry, both containing the :kconfig:option:`CONFIG_SAMPLE_SPI_MSG` string.

Reviewer 2: ```suggestion
#. Observe that the log shows a master "sent" entry and a slave "received" entry, both containing the string set by the :kconfig:option:`CONFIG_SAMPLE_SPI_MSG` Kconfig option.
```

Author: fixed

.. _bm_spi_sample:

SPI
###

.. contents::
   :local:
   :depth: 2

The SPI sample demonstrates how to configure and use the SPIM and SPIS peripherals with the nrfx drivers.

Requirements
************

The sample supports the following development kits:

.. tabs::

   .. group-tab:: Simple board variants

      The following board variants do **not** have DFU capabilities:

      .. include:: /includes/supported_boards_all_non-mcuboot_variants_s115.txt

      .. include:: /includes/supported_boards_all_non-mcuboot_variants_s145.txt

   .. group-tab:: MCUboot board variants

      The following board variants have DFU capabilities:

      .. include:: /includes/supported_boards_all_mcuboot_variants_s115.txt

      .. include:: /includes/supported_boards_all_mcuboot_variants_s145.txt

> The sample requires the following pin wiring:

Reviewer 3: ```suggestion
      .. include:: /includes/supported_boards_all_mcuboot_variants_s145.txt

Wiring
*******

The sample requires the following pin wiring:
```

Author: fixed

Single-device loopback setup:
   Wire the four SPIM pins to the SPIS pins on the same board, matching each signal by name (SCK→SCK, MOSI→MOSI, MISO→MISO, CSN→CSN).

Two-device setup:
   Wire the four SPIM pins on device 1 to the SPIS pins on device 2, matching each signal by name (SCK→SCK, MOSI→MOSI, MISO→MISO, CSN→CSN).
   Optionally, wire the four SPIS pins on device 1 to the SPIM pins on device 2 to test SPIM and SPIS on both devices.

>    .. note::

Reviewer 3: If I undrestand correctly, this note applies to both wiring configurations, so it should not be indented inside the second point.

Author: No, the indentation is correct, the note only applies to two-device, not single-device loopback setup.

In loopback, SPIM and SPIS are on the same chip, so they share one clock. The moment SPIM turns that clock on to send data, it's already on for SPIS as well, both sides wake up together because they're really the same device. There's no waiting needed since nothing was asleep to begin with.

In the two-device setup, each board manages its own power and clock separately. The receiving board can be asleep with its clock off, and the sending board has no way of knowing this, it can't see inside the other board. If data is sent right away, the receiving board might still be waking up and miss the start of the message.

That's what the delay is for. After the sender raises chip select, it waits briefly before sending real data, giving the other board time to fully wake up so it doesn't miss anything. This only matters in the two-device case, since that's the only setup where one side can be asleep while the other is talking.

Reviewer 2: I think @b-gent has a point. Yes the delay is only required in a two device setup where the SPIS requires extra wakeup time, though we do not know that at compile time. Hence the extra delay is present also in one device setup. So I think it makes sense to reduce the indentation level.  I'm also not sure whether we need to mention the alternative of keeping the high-frequency clock always on on the nRF54l when acting as a SPI device.

Author: Ok, I removed the indentation and changed the text

      Alternatively, the target can be configured to keep its high-frequency clock always on, which removes the need for the delay but increases power consumption.
      For more information, see the SPIS peripheral and electrical specification chapters in the product specification of your SoC.

.. include:: /includes/spi_board_connections.txt

.. note:: Board-specific behavior

   * **nRF54L15 DK** — The SPI controller pins overlap with **LED 3** (P1.14, flickers during transfers) and **Button 0** (P1.13, do not press during transfers).
   * **nRF54LV10 DK** — The SPI target pins (P0.00-P0.03) are shared with one of the debugger's virtual serial ports.
     Before running the sample, open the `Board Configurator`_ app in `nRF Connect for Desktop`_ and disable the **Connect port VCOM** entry that is mapped to pins **P0.00-P0.03**, to release these pins from the debugger.
     Leave the other **Connect port VCOM** entry (mapped to pins **P1.04-P1.07**) enabled, since it is the virtual serial port used to read the sample's log output.

> Overview

Reviewer 3: Move Overview before Wiring.

(as per sample template https://github.com/nrfconnect/sdk-nrf/blob/main/doc/nrf/templates/sample_README.rst#wiring)

Author: fixed

********

The sample initializes the **SPIM** and **SPIS** instances with the pins configured in the :file:`board-config.h` header.
When **Button 2** is pressed, a configurable string (:kconfig:option:`CONFIG_SAMPLE_SPI_MSG`) is sent from the local SPI controller to a connected SPI target.
The local SPI target receives data from a connected SPI controller, logs the received data and toggles **LED 2**.

User interface
**************

LED 0:
   Lit when the device is initialized.

LED 2:
   Toggles when the SPIS peripheral completes a reception.

Button 2:
   Send string from the local SPIM to a connected SPI target.

> 

Reviewer 3: Optionally, a 'Configuration' section can be added here with an explanation of the Kconfig options mentioned in this readme.

Author: Agree, added 'Configuration' section now, please take a look

Building and running
********************

This sample can be found under :file:`samples/peripherals/spi/` in the |BMshort| folder structure.

For details on how to create, configure, and program a sample, see :ref:`getting_started_with_the_samples`.

> Testing

Reviewer 3: 1. We need to mention to complete the steps in 'Wiring' first, depending on the selected config.
2. Is the Testing exactly the same for each wiring config? We could have tabs for the two different options.

Suggestion for review (with some AI assistance):

```suggestion
Testing
=======

.. tabs::

   .. group-tab:: Single-device loopback

      Test the sample on one development kit, with its SPI controller wired to its own SPI target.

      1. Wire the SPIM pins to the SPIS pins on the same kit, as described in `Wiring`_.
      #. Compile and program the application.
      #. Connect to the kit with a terminal emulator (for example, the `Serial Terminal app`_).
      #. Observe that the ``SPI sample initialized`` message is printed.
      #. Press **Button 2**.
      #. Observe that the terminal prints ``Message sent`` followed by the string set by the :kconfig:option:`CONFIG_SAMPLE_SPI_MSG` Kconfig option (``Hello World!`` by default).
      #. Observe that the terminal prints ``Message received``, the number of received bytes, and the same string.
      #. Observe that **LED 2** toggles.

   .. group-tab:: Two-device setup

      Test the sample on two development kits, with the SPI controller on device 1 wired to the SPI target on device 2.

      1. Wire device 1 and device 2 as described in `Wiring`_.
      #. Compile and program the application on both kits.
      #. Connect to both kits with a terminal emulator (for example, the `Serial Terminal app`_).
         In the following steps, these are referred to as the **device 1 terminal** and the **device 2 terminal**.
      #. Observe that both terminals print the ``SPI sample initialized`` message.
      #. Press **Button 2** on device 1.
      #. Observe that the device 1 terminal prints ``Message sent`` followed by the string set by the :kconfig:option:`CONFIG_SAMPLE_SPI_MSG` Kconfig option (``Hello World!`` by default).
      #. Observe that the device 2 terminal prints ``Message received``, the number of received bytes, and the same string.
      #. Observe that **LED 2** toggles on device 2.
      #. If you also wired the second data path, press **Button 2** on device 2 and repeat the previous three checks with the roles of the two devices reversed.
```

Author: Yes, the testing steps are the same for both wiring setups. That said, separating them into tabs makes more sense and would align with other samples like `BLE_NUS`. Your suggested change looks good, I applied it with some minor modifications, please take a look

Reviewer: Looks good!
